File: inputmodule-control/src/c1minimal.rs
use clap::Parser;

#[derive(Clone, Copy, Debug, PartialEq, clap::ValueEnum)]
pub enum Color {
    White,
    Black,
    Red,
    Green,
    Blue,
    Yellow,
    Cyan,
    Purple,
}

#[derive(Parser, Debug)]
#[command(arg_required_else_help = true)]
pub struct C1MinimalSubcommand {
    /// Set sleep status or get, if no value provided
    #[arg(long)]
    pub sleeping: Option<Option<bool>>,

    /// Jump to the bootloader
    #[arg(long)]
    pub bootloader: bool,

    /// Crash the firmware (TESTING ONLY!)
    #[arg(long)]
    pub panic: bool,

    /// Serial device, like /dev/ttyACM0 or COM0
    #[arg(long)]
    pub serial_dev: Option<String>,

    /// Get the device version
    #[arg(short, long)]
    pub version: bool,

    /// Set color
    // TODO: Allow getting current state
    #[arg(long)]
    #[clap(value_enum)]
    pub set_color: Option<Color>,
}

File: python/inputmodule/inputmodule/c1minimal.py
from inputmodule.inputmodule import send_command, CommandVals

RGB_COLORS = ["white", "black", "red", "green",
              "blue", "cyan", "yellow", "purple"]


def get_color(dev):
    res = send_command(dev, CommandVals.SetColor, with_response=True)
    return (int(res[0]), int(res[1]), int(res[2]))


def set_color(dev, color):
    rgb = None
    if color == "white":
        rgb = [0xFF, 0xFF, 0xFF]
    elif color == "black":
        rgb = [0x00, 0x00, 0x00]
    elif color == "red":
        rgb = [0xFF, 0x00, 0x00]
    elif color == "green":
        rgb = [0x00, 0xFF, 0x00]
    elif color == "blue":
        rgb = [0x00, 0x00, 0xFF]
    elif color == "yellow":
        rgb = [0xFF, 0xFF, 0x00]
    elif color == "cyan":
        rgb = [0x00, 0xFF, 0xFF]
    elif color == "purple":
        rgb = [0xFF, 0x00, 0xFF]
    else:
        print(f"Unknown color: {color}")
        return

    if rgb:
        send_command(dev, CommandVals.SetColor, rgb)
